lib/icinga/dependency.hpp

```diff
@@ -222,6 +222,42 @@ class DependencyGroup final : public SharedObject
 	static RegistryType m_Registry;
 };
 
+/**
+ * Helper class to evaluate the reachability of checkables and state of dependency groups.
+ *
+ * This class is used for implementing Checkable::IsReachable() and DependencyGroup::GetState().
+ * For this, both methods call each other, traversing the dependency graph recursively. In order
+ * to achieve linear runtime in the graph size, the class internally caches state information
+ * (otherwise, evaluating the state of the same checkable multiple times can result in exponential
+ * worst-case complexity). Because of this cached information is not invalidated, the object is
+ * intended to be short-lived.
+ */
+class DependencyStateChecker
+{
+public:
+	explicit DependencyStateChecker(DependencyType dt);
+
+	DependencyStateChecker(const DependencyStateChecker&) = delete;
+	DependencyStateChecker(DependencyStateChecker&&) = delete;
+	DependencyStateChecker& operator=(const DependencyStateChecker&) = delete;
+	DependencyStateChecker& operator=(DependencyStateChecker&&) = delete;
+
+	bool IsReachable(Checkable::ConstPtr checkable, int rstack = 0);
+	DependencyGroup::State GetState(const DependencyGroup::Ptr& group, const Checkable* child, int rstack = 0);
+
+private:
+	/**
+	 * The maximum number of allowed dependency recursion levels.
+	 *
+	 * This is a subjective limit how deep the dependency tree should be allowed to go, as anything beyond this level
+	 * is just madness and will likely result in a stack overflow or other undefined behavior.
+	 */
+	static constexpr int m_MaxDependencyRecursionLevel{256};
```

Comment:
We now have different versions of this. Wouldn't it be appropriate to put this somewhere publicly accessible, so that we don't have to maintain x versions of that very same constant.